--- a/src/unix/stream.c
+++ b/src/unix/stream.c
@@ -515,6 +515,7 @@
   assert(!(stream->flags & UV_HANDLE_CLOSING));
 
   fd = uv__stream_fd(stream);
+again:
   err = uv__accept(fd);
 
   if (err == UV_EMFILE || err == UV_ENFILE)
@@ -526,9 +527,16 @@
   stream->accepted_fd = err;
   stream->connection_cb(stream, 0);
 
-  if (stream->accepted_fd != -1)
+  if (stream->flags & UV_HANDLE_CLOSING)
+    return;
+
+  if (stream->accepted_fd != -1) {
     /* The user hasn't yet accepted called uv_accept() */
     uv__io_stop(loop, &stream->io_watcher, POLLIN);
+    return;
+  }
+
+  goto again;
 }
 
 
