diff --git a/mir-gen-aarch64.c b/mir-gen-aarch64.c
index d27f071..906b88f 100644
--- a/mir-gen-aarch64.c
+++ b/mir-gen-aarch64.c
@@ -2443,11 +2443,10 @@ static int target_memory_ok_p (gen_ctx_t gen_ctx, MIR_op_t *op_ref) {
   gen_assert (op_ref->mode == MIR_OP_VAR_MEM);
   MIR_context_t ctx = gen_ctx->ctx;
   size_t size = _MIR_type_size (ctx, op_ref->u.var_mem.type);
-  int scale = gen_int_log2 ((int64_t) size);
+  int scale = (int) size;
 
   if (op_ref->u.var_mem.disp == 0
-      && ((op_ref->u.var_mem.index == MIR_NON_VAR || op_ref->u.var_mem.scale == 1
-           || op_ref->u.var_mem.scale == scale)))
+      && (op_ref->u.var_mem.index == MIR_NON_VAR || op_ref->u.var_mem.scale == 1))
     return TRUE;
   if (op_ref->u.var_mem.index == MIR_NON_VAR && op_ref->u.var_mem.disp >= 0
       && op_ref->u.var_mem.disp % scale == 0 && op_ref->u.var_mem.disp / scale < (1 << 12))
