diff --git a/src/alloc.c b/src/alloc.c
index 6a9adf7..2d3a5ab 100644
--- a/src/alloc.c
+++ b/src/alloc.c
@@ -63,7 +63,7 @@ void tlsuv_set_allocator(uv_malloc_func malloc_f,
     ALLOC.free_f = free_f;
 
     uv_replace_allocator(malloc_f, realloc_f, calloc_f, free_f);
-#if USE_OPENSSL
+#if USE_OPENSSL && !defined(OPENSSL_IS_BORINGSSL)
     CRYPTO_set_mem_functions(crypto_malloc, crypto_realloc, crypto_free);
 #endif
 
diff --git a/src/win32/win32_keychain.c b/src/win32/win32_keychain.c
index 77378fc..d310eb3 100644
--- a/src/win32/win32_keychain.c
+++ b/src/win32/win32_keychain.c
@@ -18,6 +18,7 @@
 #include "../alloc.h"
 
 #if USE_OPENSSL
+#include <openssl/bn.h>
 #include <openssl/ecdsa.h>
 #endif
 
