--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -36,7 +36,9 @@
 endif ()
 
 set(supported_tls_libs openssl mbedtls win32crypto)
-set(TLSUV_TLSLIB ${default_crypto} CACHE STRING "TLS implementation library ${supported_tls_libs}")
+if(NOT DEFINED TLSUV_TLSLIB)
+    set(TLSUV_TLSLIB ${default_crypto})
+endif()
 
 if(NOT TLSUV_TLSLIB IN_LIST supported_tls_libs)
     message(FATAL_ERROR "`${TLSUV_TLSLIB}' TLS library is not supported")
@@ -94,7 +96,13 @@
     )
 endif (TLSUV_HTTP)
 
-add_subdirectory(src/${TLSUV_TLSLIB})
+if(TLSUV_TLSLIB STREQUAL "openssl")
+    add_subdirectory(src/openssl)
+elseif(TLSUV_TLSLIB STREQUAL "mbedtls")
+    add_subdirectory(src/mbedtls)
+elseif(TLSUV_TLSLIB STREQUAL "win32crypto")
+    add_subdirectory(src/win32crypto)
+endif()
 
 if (TARGET libuv::uv)
     message(NOTICE "upstream project set libuv target")
