diff --git a/src/http.c b/src/http.c
--- a/src/http.c
+++ b/src/http.c
@@ -1045,6 +1045,2 @@ void tlsuv_http_req_end(tlsuv_http_req_t *req) {
 int tlsuv_http_req_data(tlsuv_http_req_t *req, const char *body, size_t bodylen, tlsuv_http_body_cb cb) {
-    if (strcmp(req->method, "POST") != 0 && strcmp(req->method, "PUT") != 0) {
-        return UV_EINVAL;
-    }
-
     if (req->state > headers_sent) {
diff --git a/src/http_req.c b/src/http_req.c
--- a/src/http_req.c
+++ b/src/http_req.c
@@ -280,7 +280,8 @@ ssize_t http_req_write(tlsuv_http_req_t* req, char* buf, size_t maxlen) {
     if (strcmp(req->method, "POST") == 0 ||
         strcmp(req->method, "PUT") == 0 ||
-        strcmp(req->method, "PATCH") == 0) {
+        strcmp(req->method, "PATCH") == 0 ||
+        req->req_body != NULL) {
         if (!req->req_chunked && req->req_body_size == -1) {
             size_t req_len = 0;
             struct body_chunk_s* chunk = req->req_body;
             while (chunk != NULL) {
