diff --git a/core/shared/platform/windows/platform_internal.h b/core/shared/platform/windows/platform_internal.h
index 6cf8f4b..d35cc99 100644
--- a/core/shared/platform/windows/platform_internal.h
+++ b/core/shared/platform/windows/platform_internal.h
@@ -98,7 +98,11 @@ os_mem_commit(void *ptr, size_t size, int flags);
 void
 os_mem_decommit(void *ptr, size_t size);
 
+#if defined(__clang__)
+#define os_thread_local_attribute __thread
+#else
 #define os_thread_local_attribute __declspec(thread)
+#endif
 
 #define strncasecmp _strnicmp
 #define strcasecmp _stricmp
